### ISSUE TITLE
write_to_file generates thumbnail and exif data

### ISSUE THREAD

**Issue by tozz** [libvips/ruby-vips#147]
When using `write_to_file` ruby-vips (or vips) automatically creates a lot of metadata, the exif data isn't really a problem (but it would be nice to only have the acutally exif data in the image be the one used for generating `exif-data`) but rather the thumbnail being generated, this is quite a lot of data that probably won't get used in a lot of use cases.

This is what `get_fields` looks like when I've removed all the fields I don't need (including `jpeg-thumbnail-data`).
```
[
    [ 0] "width",
    [ 1] "height",
    [ 2] "bands",
    [ 3] "format",
    [ 4] "coding",
    [ 5] "interpretation",
    [ 6] "xoffset",
    [ 7] "yoffset",
    [ 8] "xres",
    [ 9] "yres",
    [10] "filename",
    [11] "vips-loader",
    [12] "jpeg-multiscan",
    [13] "exif-data",
    [14] "icc-profile-data",
    [15] "resolution-unit",
    [16] "orientation",
    [17] "vips-sequential"
]
```

This is how that same file looks when saving it and reopening
```
[
    [ 0] "width",
    [ 1] "height",
    [ 2] "bands",
    [ 3] "format",
    [ 4] "coding",
    [ 5] "interpretation",
    [ 6] "xoffset",
    [ 7] "yoffset",
    [ 8] "xres",
    [ 9] "yres",
    [10] "filename",
    [11] "vips-loader",
    [12] "jpeg-multiscan",
    [13] "exif-data",
    [14] "icc-profile-data",
    [15] "resolution-unit",
    [16] "exif-ifd0-Orientation",
    [17] "exif-ifd0-XResolution",
    [18] "exif-ifd0-YResolution",
    [19] "exif-ifd0-ResolutionUnit",
    [20] "exif-ifd1-Compression",
    [21] "exif-ifd1-XResolution",
    [22] "exif-ifd1-YResolution",
    [23] "exif-ifd1-ResolutionUnit",
    [24] "exif-ifd2-PixelXDimension",
    [25] "exif-ifd2-PixelYDimension",
    [26] "exif-ifd2-ExifVersion",
    [27] "exif-ifd2-FlashPixVersion",
    [28] "exif-ifd2-ColorSpace",
    [29] "jpeg-thumbnail-data",
    [30] "orientation",
    [31] "vips-sequential"
]
```

Supplying `strip: true` to `write_to_file` removes too much (I need to keep ICC and EXIF copyright where available).

Any way to not have automatically generated thumbnails when saving? Or is it a case of removing the field actually not removing the binary data?

**Comment by jcupitt** [comment]
Hello, you need to remove `exif-data-block` as well.

When libvips loads a JPG image, it extracts the EXIF block and attaches it as a chunk of binary data. It then uses libexif to parse the EXIF data and generates a set of extra string type metadata items for each key it discovers. These are named as `exif-ifd0-XResolution', etc. There are some obvious rules for turning binary data, like rationals, for example, into strings.

User code can change these extra fields, for example setting new values, or creating new tags, as long as they follow the syntax rules for exif metadata.

On write, libvips looks for the `exif-data-block` binary tag and uses libexif to parse it again into a tree structure. If there's no exif data block, it generates the minimal set required by the exif standard. It then walks the image metadata looking for special `exif-` items, matches each up with the right place in the tree, and updates, creates or removes the tree value there.

Once the exif structure has been rebuilt and updated, it uses libexif to serialise it back into a new exif block and writes that to the output JPG image.

So ... your thumbnail was being recreated from the binary exif data block that libvips was carrying around. Delete that too and it should work. 

I guess it should probably delete the exif thumbnail if the thumbnail data has been removed.

**Comment by tozz** [comment]
Removing the `exif-data` block will however invalidate any `exif-` field in the image, for example, if I keep `exif-ifd0-Copyright` and remove `exif-data` the copyright will be `[None] (Photographer) - [None] (Editor)`, probably due to some missing reference (I'm guessing here).
The thumbnail does get removed though.
I know vips is not really an exif-manipulation software, but would it be possible to acutally maintain some of the data while wiping the rest?

**Comment by jcupitt** [comment]
Huh funny, I'd think it would keep copyright. This looks like a bug, I'll see if I can fix it.

**Comment by jcupitt** [comment]
Sorry, this fell off my radar :( 

I've added a thing to 8.7 to remove the thumbnail from the EXIF block if the user removes the tag. So this now works:

```ruby
image = Vips::Image.new_from_file ARGV[0]
image.remove 'jpeg-thumbnail-data'
image.write_to_file ARGV[1]
```

The thing you noticed with the copyright tag appears to be a libexif feature. This tag is supposed to be split into two fields with NULL characters, the first giving the name of the photographer, and the second the editor. I'd guess that the string you have in the image does not follow this pattern, so libexif gets confused. 

Does this resolve your issue?

**Comment by janko-m** [comment]
I'm not able to remove EXIF data by removing `exif-data`:

```rb
require "vips"
require "open-uri"

file = open("https://upload.wikimedia.org/wikipedia/commons/3/36/Hopetoun_falls.jpg")

image = Vips::Image.new_from_file(file.path)
image.remove("exif-data")

saved_image = Vips::Image.new_from_buffer("#{image.write_to_buffer(".jpg")}", "")

p saved_image.get_fields.grep(/exif/)
```
```
["exif-data", "exif-ifd0-Make", "exif-ifd0-Model", "exif-ifd0-Orientation", ...]
```

The only field that I noticed it removed was `jpeg-thumbnail-data`. I'm using libvips 8.6.3 and ruby-vips 2.0.9.

**Comment by jcupitt** [comment]
Hello, when the jpeg writer writes, it always attaches some exif, unless you use the `strip` option. 

```
john@kiwi:~/pics$ irb
irb(main):001:0> require 'vips'
=> true
irb(main):002:0> x = Vips::Image.new_from_file "IMG_0073.JPG"
=> #<Image 5472x3648 uchar, 3 bands, srgb>
irb(main):003:0> x.get_fields.length
=> 65
irb(main):004:0> x.remove "exif-data"
=> nil
irb(main):005:0> x.write_to_file("x.jpg")
vips warning: exif: bad exif meta "exif-ifd3-GPSVersionID"
=> nil
irb(main):006:0> y = Vips::Image.new_from_file "x.jpg"
=> #<Image 5472x3648 uchar, 3 bands, srgb>
irb(main):007:0> y.get_fields.length
=> 57
```

I think 57 is the minimum amount of exif the spec allows.

**Comment by jcupitt** [comment]
Oh, this gets it down a bit more:

```
john@kiwi:~/pics$ irb
irb(main):001:0> require 'vips'
=> true
irb(main):002:0> x = Vips::Image.new_from_file "IMG_0073.JPG"
=> #<Image 5472x3648 uchar, 3 bands, srgb>
irb(main):003:0> x.get_fields.each {|field| if field =~ /exif-ifd.*/ then x.remove(field) end}
=> ["width", "height", "bands", "format", "coding", "interpretation", "xoffset", "yoffset", "xres", "yres", "filename", "vips-loader", "jpeg-multiscan", "exif-data", "resolution-unit", "exif-ifd0-Make", "exif-ifd0-Model", "exif-ifd0-Orientation", "exif-ifd0-XResolution", "exif-ifd0-YResolution", "exif-ifd0-ResolutionUnit", "exif-ifd0-DateTime", "exif-ifd0-Artist", "exif-ifd0-YCbCrPositioning", "exif-ifd0-Copyright", "exif-ifd1-Compression", "exif-ifd1-XResolution", "exif-ifd1-YResolution", "exif-ifd1-ResolutionUnit", "exif-ifd2-ExposureTime", "exif-ifd2-FNumber", "exif-ifd2-ExposureProgram", "exif-ifd2-ISOSpeedRatings", "exif-ifd2-ExifVersion", "exif-ifd2-DateTimeOriginal", "exif-ifd2-DateTimeDigitized", "exif-ifd2-ComponentsConfiguration", "exif-ifd2-ShutterSpeedValue", "exif-ifd2-ApertureValue", "exif-ifd2-ExposureBiasValue", "exif-ifd2-MeteringMode", "exif-ifd2-Flash", "exif-ifd2-FocalLength", "exif-ifd2-MakerNote", "exif-ifd2-UserComment", "exif-ifd2-SubsecTime", "exif-ifd2-SubSecTimeOriginal", "exif-ifd2-SubSecTimeDigitized", "exif-ifd2-FlashPixVersion", "exif-ifd2-ColorSpace", "exif-ifd2-PixelXDimension", "exif-ifd2-PixelYDimension", "exif-ifd2-FocalPlaneXResolution", "exif-ifd2-FocalPlaneYResolution", "exif-ifd2-FocalPlaneResolutionUnit", "exif-ifd2-CustomRendered", "exif-ifd2-ExposureMode", "exif-ifd2-WhiteBalance", "exif-ifd2-SceneCaptureType", "exif-ifd3-GPSVersionID", "exif-ifd4-InteroperabilityIndex", "exif-ifd4-InteroperabilityVersion", "jpeg-thumbnail-data", "orientation", "xmp-data"]
irb(main):004:0> x.get_fields
=> ["width", "height", "bands", "format", "coding", "interpretation", "xoffset", "yoffset", "xres", "yres", "filename", "vips-loader", "jpeg-multiscan", "exif-data", "resolution-unit", "jpeg-thumbnail-data", "orientation", "xmp-data"]
irb(main):005:0> x.write_to_file "x.jpg"
=> nil
irb(main):006:0> y = Vips::Image.new_from_file "x.jpg"
=> #<Image 5472x3648 uchar, 3 bands, srgb>
irb(main):007:0> y.get_fields.length
=> 31
```

**Comment by janko-m** [comment]
@jcupitt Thanks a lot for the explanation, that makes sense 👍

**Comment by jcupitt** [comment]
OK, this should be fixed in 8.7 (which is now out, finally).

I'll close. Please open a new issue if there's still a problem.
